TanStack Start SDK (#1495)
clerk/clerk-docs: docs/references/tanstack-start/read-session-data.mdx

---
title: Read session and user data in your TanStack Start app with Clerk
description: Learn how to use Clerk's hooks and helpers to access the active session and user data in your TanStack Start application.
---

Clerk provides a set of [hooks and helpers](/docs/references/tanstack-start/overview#client-side-helpers) that you can use to access the active session and user data in your TanStack Start application. Here are examples of how to use these helpers in both the client and server side to get you started.

## Server side

### `getAuth()`

To use `getAuth` on the server side, you can import it from `@clerk/tanstack-start/server` and use it in your server functions and API Routes to access the `Auth` object.

<Tabs type="function" items={["Before load Function", "API Endpint"]}>
  <Tab>

>     ```tsx {{ prettier: false, filename: 'app/routes/example.tsx' }}

Reviewer: ```suggestion
    ```tsx {{ filename: 'app/routes/example.tsx' }}
```

      import { getAuth } from '@clerk/tanstack-start/server'
      import { createFileRoute, redirect } from '@tanstack/react-router'
      import { createServerFn } from '@tanstack/start'

      const getAuthenticatedUser = createServerFn('POST', async (_, { request }) => {
        const auth = await getAuth(request)

        if (!auth.userId) {
          throw redirect({
            to: '/sign-in',
          })
        }

        return { auth }
      })

      export const Route = createFileRoute('/example')({
        component: Example,
        beforeLoader: async () => {
          const { auth } = await getAuthenticatedUser()

          return { auth }
        },
        loader: async ({ context }) => {
          return {
            userId: context.auth.userId,
          }
        },
      })

      async function Example() {
        const { userId } = await getAuth()

        return <div>Hello, {userId}!</div>
      }
    ```
  </Tab>

  <Tab>

>     ```tsx {{ prettier: false, filename: 'app/routes/api/example.tsx' }}

Reviewer: ```suggestion
    ```tsx {{ filename: 'app/routes/api/example.tsx' }}
```

    import { getAuth } from '@clerk/nextjs/server'
    import { json } from '@tanstack/start'
    import { createAPIFileRoute } from '@tanstack/start/api'

    export const Route = createAPIFileRoute('/api/example')({
      GET: async ({ request, params }) => {
        const { userId, getToken } = await getAuth(request);

        if (!userId) {
          return json({ error: 'Unauthorized' }, { status: 401 });
        }

        return json({ userId });
      },
    })
    ```
  </Tab>
</Tabs>

## Client Side

### `useAuth()`

The `useAuth` hook is a convenient way to access the current auth state. This hook provides the minimal information needed for data-loading and helper methods to manage the current active session.

> ```tsx {{ prettier: false, filename: 'routes/example.tsx' }}

Reviewer: ```suggestion
```tsx {{ filename: 'routes/example.tsx' }}
```

  import { useAuth } from '@clerk/tanstack-start'
  import { SignedIn, UserButton, SignOutButton, SignedOut, SignInButton, SignUpButton } from '@clerk/tanstack-start'
  import { createFileRoute } from '@tanstack/react-router'

  export const Route = createFileRoute('/example')({
    component: Example,
  })

  function Example() {
    const { isLoaded, userId, sessionId, getToken } = useAuth();

    if (!isLoaded) {
      return <p>Loading...</p>
    }

    if (!userId) {
      // You could also add a redirect to the sign-in page here
      return null
    }

    return <div>Hello, {userId}!</div>
  }
```

### `useUser()`

The `useUser` hook is a convenient way to access the current user data where you need it. This hook provides the user data and helper methods to manage the current active session.

> ```tsx {{ prettier: false, filename: 'routes/example.tsx' }}

Reviewer: ```suggestion
```tsx {{ filename: 'routes/example.tsx' }}
```

---
title: Read session and user data in your TanStack Start app with Clerk
description: Learn how to use Clerk's hooks and helpers to access the active session and user data in your TanStack Start application.
---

Clerk provides a set of [hooks and helpers](/docs/references/tanstack-start/overview#client-side-helpers) that you can use to access the active session and user data in your TanStack Start application. Here are examples of how to use these helpers in both the client and server side to get you started.

## Server side

### `getAuth()`

> To use `getAuth` on the server side, you can import it from `@clerk/tanstack-start/server` and use it in your server functions and API Routes to access the `Auth` object.

Reviewer 2: ```suggestion
The [`getAuth()`](/docs/references/nextjs/get-auth){{ target: '_blank' }} helper allows you to access the [`Auth` object](/docs/references/nextjs/auth-object){{ target: '_blank' }}, which includes the current user's `userId`. You can use the `userId` to protect your routes or get the user's data.
```

<Tabs type="function" items={["Before load Function", "API Endpint"]}>
  <Tab>
    ```tsx {{ filename: 'app/routes/example.tsx' }}
    import { getAuth } from '@clerk/tanstack-start/server'
    import { createFileRoute, redirect } from '@tanstack/react-router'
    import { createServerFn } from '@tanstack/start'

    const getAuthenticatedUser = createServerFn('POST', async (_, { request }) => {
      const auth = await getAuth(request)

      if (!auth.userId) {
        throw redirect({
          to: '/sign-in',
        })
      }

      return { auth }
    })

    export const Route = createFileRoute('/example')({
      component: Example,
      beforeLoader: async () => {
        const { auth } = await getAuthenticatedUser()

        return { auth }
      },
      loader: async ({ context }) => {
        return {
          userId: context.auth.userId,
        }
      },
    })

    async function Example() {

>       const { userId } = await getAuth()

Reviewer 2: `const state = Route.useLoaderData()` ?
if you are doing `const { userId } = await getAuth()` , then why need to create `getAuthenticatedUser` and pass it to `loader` ?

can you explain this example to me

Author: Fixed!

      return <div>Hello, {userId}!</div>
    }
    ```
  </Tab>

  <Tab>
    ```ts {{ filename: 'app/routes/api/example.ts' }}
    import { getAuth } from '@clerk/nextjs/server'
    import { json } from '@tanstack/start'
    import { createAPIFileRoute } from '@tanstack/start/api'

    export const Route = createAPIFileRoute('/api/example')({
      GET: async ({ request, params }) => {
        const { userId, getToken } = await getAuth(request)

        if (!userId) {
          return json({ error: 'Unauthorized' }, { status: 401 })
        }

        return json({ userId })
      },
    })
    ```
  </Tab>
</Tabs>

## Client Side

### `useAuth()`

> The `useAuth` hook is a convenient way to access the current auth state. This hook provides the minimal information needed for data-loading and helper methods to manage the current active session.

Reviewer 2: ```suggestion
The [`useAuth()`](/docs/references/react/use-auth){{ target: '_blank' }} hook provides information about the current auth state, as well as helper methods to manage the current active session. The hook returns `userId`, which can be used to protect your routes.
```

```tsx {{ filename: 'routes/example.tsx' }}
import { useAuth } from '@clerk/tanstack-start'
import {
  SignedIn,
  UserButton,
  SignOutButton,
  SignedOut,
  SignInButton,
  SignUpButton,
} from '@clerk/tanstack-start'
import { createFileRoute } from '@tanstack/react-router'

export const Route = createFileRoute('/example')({
  component: Example,
})

function Example() {

>   const { isLoaded, userId, sessionId, getToken } = useAuth()

Reviewer 2: These aren't being used

```suggestion
  const { isLoaded, userId } = useAuth()
```

  if (!isLoaded) {
    return <p>Loading...</p>
  }

  if (!userId) {
    // You could also add a redirect to the sign-in page here
    return null
  }

  return <div>Hello, {userId}!</div>
}
```

### `useUser()`

> The `useUser` hook is a convenient way to access the current user data where you need it. This hook provides the user data and helper methods to manage the current active session.

Reviewer 2: ```suggestion
The [`useUser()`](/docs/references/react/use-user){{ target: '_blank' }} hook provides the current user's [`User`](/docs/references/javascript/user/user){{ target: '_blank' }} object, which holds all of the information for a user of your application and provides a set of methods to manage their account.
```
